pkg/controller/sidecarset/sidecarset_processor.go

```diff
@@ -0,0 +1,349 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sidecarset
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	appsv1alpha1 "github.com/openkruise/kruise/apis/apps/v1alpha1"
+	"github.com/openkruise/kruise/pkg/sidecarcontrol"
+	"github.com/openkruise/kruise/pkg/util"
+	"github.com/openkruise/kruise/pkg/util/expectations"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type Processor struct {
+	Client             client.Client
+	recorder           record.EventRecorder
+	updateExpectations expectations.UpdateExpectations
+}
+
+func NewSidecarSetProcessor(cli client.Client, expectations expectations.UpdateExpectations, rec record.EventRecorder) *Processor {
+	return &Processor{
+		Client:             cli,
+		updateExpectations: expectations,
+		recorder:           rec,
+	}
+}
+
+func (p *Processor) UpdateSidecarSet(sidecarSet *appsv1alpha1.SidecarSet) (reconcile.Result, error) {
+	// SidecarSet upgrade strategy type is NotUpdate
+	if !isSidecarSetNotUpdate(sidecarSet) {
+		return reconcile.Result{}, nil
+	}
+	control := sidecarcontrol.New(sidecarSet)
+
+	// 1. get matching pods with the sidecarSet
+	pods, err := p.getMatchingPods(sidecarSet)
+	if err != nil {
+		klog.Errorf("sidecarSet get matching pods error, err: %v, name: %s", err, sidecarSet.Name)
+		return reconcile.Result{}, err
+	}
+
+	// 2. calculate SidecarSet status based on pods
+	status := calculateStatus(control, pods)
+	//update sidecarSet status in store
+	if err := p.updateSidecarSetStatus(sidecarSet, status); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// in case of informer cache latency
+	for _, pod := range pods {
+		p.updateExpectations.ObserveUpdated(sidecarSet.Name, sidecarcontrol.GetSidecarSetRevision(sidecarSet), pod)
+	}
+	allUpdated, _, inflightPods := p.updateExpectations.SatisfiedExpectations(sidecarSet.Name, sidecarcontrol.GetSidecarSetRevision(sidecarSet))
+	if !allUpdated {
+		klog.V(3).Infof("sidecarset %s matched pods has some update in flight: %v, will sync later", sidecarSet.Name, inflightPods)
+		return reconcile.Result{RequeueAfter: time.Second}, nil
+	}
+
+	// 4. sidecarset already updates all matched pods, then return
+	if isSidecarSetUpdateFinish(status) {
+		klog.V(3).Infof("sidecarSet update pod finished, name: %s", sidecarSet.Name)
+		return reconcile.Result{}, nil
+	}
+
+	// 5. Paused indicates that the SidecarSet is paused to update matched pods
+	if sidecarSet.Spec.Strategy.Paused {
+		klog.V(3).Infof("sidecarSet is paused, name: %s", sidecarSet.Name)
+		return reconcile.Result{}, nil
+	}
+
+	// 6. if sidecarset can upgrade pods
+	if !control.IsSidecarSetCanUpgrade(pods[0]) {
```

Comment:
Is it better to move this into `updatePods` loop before each Pod to update?

Reply:
In kubernetes cluster, when inplace update pod, only fields such as image can be updated for the container.  It is to determine whether there are other fields that have been modified here. I will try to implement it in a way that is easier to understand.

Reply:
done